Puzzle map indices

iimarckus opened the pull request:
These are the functions that turn the player’s map and map coordinate (𝑥 ranging from ⟦0, 4⟧ and 𝑦 ⟦0, 2⟧) into a puzzle ID.

mid-kid commented on source/bank_01.asm:
```diff
@@ -31,6 +32,38 @@ level_name_textbox_clear:
+SECTION "calc_town_map_level_index", ROMX[$61FA], BANK[$01]
+calc_town_map_level_index::
+    ld a, [w_town_map_zone]
+    cp a, 10
```
```suggestion
    cp 10
```

mid-kid replied:
Also, if the zones have indices, please create constants for them - they have proper names in-game.

mid-kid commented on source/bank_01.asm:
```diff
@@ -31,6 +32,38 @@ level_name_textbox_clear:
+    add c
+    ld c, a
+    ld a, [w_town_map_zone]
+    xor a, $FF
```
```suggestion
    xor $FF
```

Rangi42 replied:
Also it's generally lowercase `$ff`.

mid-kid commented on source/bank_01.asm:
```diff
@@ -31,6 +32,38 @@ level_name_textbox_clear:
+    xor a, $FF
+    inc a
+    add c
+    add a, LEVEL_DIGLETT
```
```suggestion
    add LEVEL_DIGLETT
```

mid-kid commented on source/bank_20.asm:
```diff
@@ -0,0 +1,50 @@
+    ld a, [hl]
+    jp farcall_ret
+
+.zone_4_puzzles
```
```suggestion
.zone_4_puzzles:
```
We use colons for anything that isn't local jumps, so standalone data applies.

mid-kid commented on source/bank_20.asm:
```diff
@@ -0,0 +1,50 @@
+    xor a, $FF
+    inc a
+    add c
+    add a, LEVEL_SAFARI_HAUNTER
```
```suggestion
    add LEVEL_SAFARI_HAUNTER
```

mid-kid commented on source/bank_20.asm:
```diff
@@ -0,0 +1,50 @@
+    add c
+    ld c, a
+    ld a, [w_safari_map_zone]
+    xor a, $FF
```
```suggestion
    xor $FF
```

mid-kid commented on source/bank_20.asm:
```diff
@@ -0,0 +1,50 @@
+SECTION "calc_safari_map_level_index", ROMX[$46B5], BANK[$20]
+calc_safari_map_level_index::
+    ld a, [w_safari_map_zone]
+    cp a, 3
```
```suggestion
    cp 3
```

Rangi42 commented on shim.sym:
```diff
@@ -218,6 +218,14 @@
 01:d6c8 w_d6c8
 01:d6c9 w_d6c9
 01:d6ca w_d6ca
+01:d6d4 w_town_map_zone ; db
```
If these are the levels (1-10 for Town Map, 1-4 for Safari Zone), they should say `level`, not `zone`, for consistency and to avoid confusion with the Safari Zone. (Thus `w_town_map_level` and `w_safari_map_level`, or `w_safari_zone_level`.)

iimarckus replied:
The reason I called them zones is that “level” is already used to refer to individual puzzles.

iimarckus replied:
Alternatively we could rename “levels” to “puzzles”, `level_names.asm` to `puzzle_names.asm`, etc., and repurpose “level” as you describe. I didn’t really like calling them “zones” outside of the Safari Map either.
![safari-map](https://user-images.githubusercontent.com/1767429/93917392-2ef9c100-fcc8-11ea-854b-af7176e1b42f.png)
![town-map](https://user-images.githubusercontent.com/1767429/93917406-3325de80-fcc8-11ea-953b-827b7fafcf31.png)

Rangi42 replied:
Their graphics are marked "LV1", "LV2", etc, even in the Safari Zone, so yeah, `level` for those and `puzzle` for the individual puzzles.

iimarckus commented:
Renamed levels to puzzles in existing code.

Dropped implicit `a` (blame BGB).

In new code:

* lowercased $FF
* referred to maps consistently (Easy Picross, Town Map, Safari Map)
* referred to puzzles, not levels
* referred to levels, not zones

mid-kid commented:
Can you do a really quick `sed -i -e 's/=/EQU/' include/{level,puzzle}_constants.inc`?

iimarckus commented:
Done.

Rangi42 commented on source/bank_20.asm:
```diff
@@ -0,0 +1,51 @@
+    xor $ff
+    inc a
+    add c
+    add PUZZLE_SAFARI_HAUNTER
```
There should be a `FIRST_SAFARI_ZONE_PUZZLE` constant for this, like how pokecrystal has `JOHTO_POKEMON EQU const_value` in pokemon_constants.asm or `KANTO_LANDMARK EQU const_value` in landmark_constants.asm.

Rangi42 commented on source/bank_01.asm:
```diff
@@ -31,6 +33,38 @@ level_name_textbox_clear:
+    xor $ff
+    inc a
+    add c
+    add PUZZLE_DIGLETT
```
Likewise a `FIRST_TOWN_MAP_PUZZLE` constant for this (the previous 15 are Easy Picross)

mid-kid commented:
@Rangi42 Sorry, I've already merged this, but feel free to create those constants (and maybe use `const` to define them).
